Add TO Go cdns/health and cdns/{name}/health

rob05c opened the pull request:
Rewrites `/cdns/health` and `/cdns/{{name}}/health` to Go. Closes #3778 and closes #3779 .

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1623/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/1624/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/1900/
Test PASSed.

dneuman64 commented on lib/go-tc/enum.go:
```diff
@@ -53,6 +53,8 @@ type DeliveryServiceName string
 // CacheType is the type (or tier) of a CDN cache.
 type CacheType string
 
+const MonitorTypeName = "RASCAL"
```
Do we think this should be a config?  I have seen several times now where the type "RASCAL" was hard-coded and we had to change our TM type.  IMO you should be able to make the type whatever you want and it should work.

rob05c replied:
I think it should definitely be a config, once we've moved everything to Go. But right now, it's still hard-coded all over Perl. 

I don't think we should make a config yet, which would make people think they could change it, when it'd still break to be anything but "RASCAL".

But definitely +100 when it's all Go, and we can set one variable. I've wanted to change our db "RASCAL" type to something like "monitor" for ages.

dneuman64 commented:
I would expect a heathtest.go to be a part of this PR is there a reason it is not included?

rob05c commented:
Not sure what you mean. What would be in healthtest.go?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/1998/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2039/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2114/
Test FAILed.

ocket8888 commented:
This endpoint has already been rewritten to go, so I think this PR should be closed.

rob05c commented:
This hasn't been rewritten in Go, except for this PR. It's still proxtying to Perl https://github.com/apache/trafficcontrol/blob/master/traffic_ops/traffic_ops_golang/routing/routes.go#L126

ocket8888 commented:
Oh, lol, my bad. I forget that some of those routes exist for explicit routing to the proxy. If you wanna rebase for conflicts, I'd be happy to review this.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4163/
Test PASSed.

rob05c commented on traffic_ops/traffic_ops_golang/util/monitorhlp/monitorhlp.go:
```diff
@@ -0,0 +1,116 @@
+	return monitors, nil
+}
+
+func GetCRStates(monitorFQDN string, client *http.Client) (tc.CRStates, error) {
```
Yes, they were copy-pasted, because the PRs were open at the same time, with the intention of combing them after this was merged to master. Good catch. I'll do that now.

rob05c replied:
Done. Also with `GetCRConfig`.

rob05c commented:
>also should we be adding tests?

This endpoint calls out to Traffic Monitor. The "API Test" framework doesn't currently support Monitors. It's also not really possible to unit test; we could mock everything, but it wouldn't be testing much of anything at that point (which is why we've been doing "API Tests" for TO).

We should definitely test this (and other Monitor endpoints) when the API Tests support it. But it'll be a nontrivial amount of work to add that support.

mitchell852 commented:
@rob05c - i believe this PR is going to have the same problem that was addressed in #3927

mhoppa commented:
Yes it will have the same problem

mitchell852 commented:
@rob05c - i think this PR needs the changes that were made in #3927

rob05c commented:
I'd like to look into #3927 first. Not sure when I'll have time. 

It shouldn't be necessary, HTTP/2 is backwards-compatible by design. If a client tries to use H2, and a server doesn't support it, it automatically falls back to H1. If it doesn't, the Server is violating HTTP/1.1. So, it shouldn't ever be necessary to forcibly disable it. I'd like to at least understand why it's necessary.

rob05c commented:
Ok, I investigated, determined Go 1.13 http.Transport fundamentally doesn't support H2 Proxies. I added the code from #3927 and a comment to that effect. See https://github.com/golang/go/issues/26479 "We only support http1 proxies currently." 

Rebased; manually tested again, still works as expected, still matches Perl.
